### PR TITLE
hw5: hooks

### PR DESCRIPTION
**Pull request by kamiour**
### Implemented the following hooks:
- useState
- useEffect
- useContext
- useId
- useCallback
- useMemo

#### All functionality of the app except forms now works with mocked data.

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -6,16 +6,39 @@ import Main from './components/Main/Main';
...
+interface SelectedMovieState {
+  selectedMovie: Movie | null;
+  setSelectedMovie: Dispatch<SetStateAction<Movie | null>>;
+}
+
+export const SelectedMovieContext = React.createContext<SelectedMovieState>({ selectedMovie: null, setSelectedMovie: () => {} });
+export const FetchedMoviesContext = React.createContext<UseFetchMoviesResult>([
+  { fetchedMovies: [], isError: false, isLoading: false, queryParams: {} },
+  () => {},
+]);
```
Comment (Twixs) [inline src/App.tsx:22]: Unusual to see several contexts (but it's good that you thought about logic separation in Redux-style) :D

Please move context creation in separate folder/file

Comment (Twixs) [inline src/App.tsx:22]: Unusual to see several contexts (but it's good that you thought about logic separation in Redux-style) :D

Please move context creation in separate folder/file

File: src/components/MovieListCard/MovieListCard.tsx [hunk 2]
```diff
@@ -29,40 +33,48 @@ function MoviesListCard({ movie }: MoviesListCardProps) {
...
 
-  function handleMovieEdit(formValue: Partial<Movie>) {
+  const handleMovieEdit = useCallback((formValue: Partial<Movie>) => {
     // edit request
     console.log(formValue);
-  }
+  }, []);
+
+  const closeEditMovieModal = useCallback(() => setMovieToEdit(null), []);
+  const closeDeleteMovieModal = useCallback(() => setMovieToDelete(null), []);
```
Comment (Twixs) [inline src/components/MovieListCard/MovieListCard.tsx:65]: useCallback is redundant here. Please take a look at this article https://kentcdodds.com/blog/usememo-and-usecallback

Comment (Twixs) [inline src/components/MovieListCard/MovieListCard.tsx:65]: useCallback is redundant here. Please take a look at this article https://kentcdodds.com/blog/usememo-and-usecallback

File: src/hooks/useFetchMovies.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+export type UseFetchMoviesResult = [
+  { fetchedMovies: Movie[]; isLoading: boolean; isError: boolean; queryParams: { [key: string]: string } },
+  React.Dispatch<React.SetStateAction<{ [key: string]: string }>>
+];
+
+export const useFetchMovies = (initialData: Movie[]): UseFetchMoviesResult => {
+  const apiUrl = 'API_URL';
+
+  const [fetchedMovies, setFetchedMovies] = useState(initialData);
+  const [queryParams, setQueryParams] = useState({}); // to react to search/sort/filter values
```
Comment (Twixs) [inline src/hooks/useFetchMovies.ts:14]: 1 - what is the point of passing initialData here?
2 - you can get queryParams from FetchedMoviesContext.

Comment (Twixs) [inline src/hooks/useFetchMovies.ts:14]: 1 - what is the point of passing initialData here?
2 - you can get queryParams from FetchedMoviesContext.

File: src/App.tsx (hunk 1 above)
Comment (kamiour) [reply src/App.tsx:22]: Thank you, moved

**Pull request merged**